x/fswap/keeper/keeper.go

```diff
@@ -1,42 +1,345 @@
+	cdc      codec.BinaryCodec
+	storeKey storetypes.StoreKey
 
-func NewKeeper(
-	cdc codec.BinaryCodec,
-	storeKey storetypes.StoreKey,
-	ak types.AccountKeeper,
-	bk types.BankKeeper,
-	config types.Config,
-) Keeper {
+	AccountKeeper
+	BankKeeper
+
+	fromDenom    string
+	toDenom      string
+	swapMultiple sdk.Int
+	swapCap      sdk.Int
+}
+
+func NewKeeper(cdc codec.BinaryCodec, storeKey storetypes.StoreKey, ak AccountKeeper, bk BankKeeper) Keeper {
 	return Keeper{
 		cdc:           cdc,
 		storeKey:      storeKey,
-		accountKeeper: ak,
-		bankKeeper:    bk,
-		config:        config,
+		AccountKeeper: ak,
+		BankKeeper:    bk,
+		fromDenom:     "",
+		toDenom:       "",
+		swapMultiple:  sdk.ZeroInt(),
+		swapCap:       sdk.ZeroInt(),
 	}
 }
 
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
+
+func (k Keeper) FswapInit(ctx sdk.Context, fswapInit types.FswapInit) error {
+	if err := fswapInit.ValidateBasic(); err != nil {
+		return err
+	}
+	if k.hasBeenInitialized(ctx) {
+		return errors.New("already initialized")
+	}
+	if err := k.setFswapInit(ctx, fswapInit); err != nil {
+		return err
+	}
+	swapped := types.Swapped{
+		OldCoinAmount: sdk.Coin{
+			Denom:  fswapInit.GetFromDenom(),
+			Amount: sdk.ZeroInt(),
+		},
+		NewCoinAmount: sdk.Coin{
+			Denom:  fswapInit.GetToDenom(),
+			Amount: sdk.ZeroInt(),
+		},
+	}
+	if err := k.setSwapped(ctx, swapped); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) Swap(ctx sdk.Context, addr sdk.AccAddress, oldCoinAmount sdk.Coin) error {
+	if ok := k.HasBalance(ctx, addr, oldCoinAmount); !ok {
+		return sdkerrors.ErrInsufficientFunds
+	}
+	fswapInit, err := k.getFswapInit(ctx)
+	if err != nil {
+		return err
+	}
+	if oldCoinAmount.GetDenom() != fswapInit.GetFromDenom() {
+		return errors.New("denom mismatch")
+	}
+
+	newAmount := oldCoinAmount.Amount.Mul(fswapInit.SwapMultiple)
+	newCoinAmount := sdk.NewCoin(fswapInit.ToDenom, newAmount)
+	if err := k.checkSwapCap(ctx, newCoinAmount); err != nil {
+		return err
+	}
+
+	moduleAddr := k.GetModuleAddress(types.ModuleName)
+	if err := k.SendCoins(ctx, addr, moduleAddr, sdk.NewCoins(oldCoinAmount)); err != nil {
+		return err
+	}
+
+	if err := k.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(oldCoinAmount)); err != nil {
+		return err
+	}
+
+	if err := k.MintCoins(ctx, types.ModuleName, sdk.NewCoins(newCoinAmount)); err != nil {
+		return err
+	}
+
+	if err := k.SendCoins(ctx, moduleAddr, addr, sdk.NewCoins(newCoinAmount)); err != nil {
+		return err
+	}
+
+	if err := k.updateSwapped(ctx, oldCoinAmount, newCoinAmount); err != nil {
+		return err
+	}
+
+	if err := ctx.EventManager().EmitTypedEvent(&types.EventSwapCoins{
+		Address:       addr.String(),
+		OldCoinAmount: oldCoinAmount,
+		NewCoinAmount: newCoinAmount,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) SwapAll(ctx sdk.Context, addr sdk.AccAddress) error {
```

Comment:
`SwapAll()` now has to get denoms as parameters.

Reply:
There are some updated code.
Please check updated code has the same problem.